skt/__init__.py

```diff
@@ -405,8 +405,11 @@ def merge_patchwork_patch(self, uri):
                           patchinfo.get("name").replace(',', ';')))
 
     def merge_patch_file(self, path):
+        if not os.path.isfile(path):
+            raise Exception("Patch %s not found" % os.path.abspath(path))
+        abspath = os.path.abspath(path)
```

Comment:
We already have many cases of normalizing paths deep in the code which need to be fixed. Let's not add another one. Instead, call `os.path.abspath` in `cmd_merge`, or better yet during configuration/command-line option parsing/loading.

The next step would be to document the function, mention that the path should be absolute, and verify that inside the function (e.g. with an `assert` and `os.path.isabs`). @veruu, can we have assertions in our Python code? What are the recommended practices?

Reply:
I think I agree with @veruu 's perspective

Reply:
> Technically we can, but assertions are usually associated with tests instead of production code.

So Python test framework authors hijacked `assert`, apparently not knowing its original purpose, and made it useless? Oh, well...

> If part X of the code base is responsible for making the path absolute, verification it works as expected should be part of tests for part X.

Writing tests checking that a function interface is observed for every invocation of that function doesn't scale. Letting the function check its arguments works better. Whether it's with an assertion, or with an if statement, doesn't really matter.

BTW, this is called "precondition checking" in "[Design by contract](https://en.wikipedia.org/wiki/Design_by_contract)" approach, which I find very effective for keeping developers' sanity, and would like to see used in our code as well.

> It is harder to debug the problem if you get an error in unrelated part of the code.

That's why we have backtraces :)

Reply:
One of disadvantages of using `assert` instead of custom `try/except` block is that you always get `AssertionError` instead of `TypeError`, `ValueError` etc, which are more appropriate in the main code base. Indeed, you can (and should) pass your own message, but the exception class stays the same, which might create some issues in unexpected places.

For example, function A should raise `ValueError` if passed value doesn't fit expected range, and `TypeError` if the input is wrong type. Function B, which calls A, might want to do something specific (eg. cleanup) based on the exception A raises. Raising same `AttributeError` for both different issues will just complicate things (and doesn't go well with *explicit is better than implicit* rule).

Another

Reply:
The explanation got a bit long so I made the relevant tl;dr bold for ease.

One of disadvantages of using assert instead of throwing exception is that you always get `AssertionError` instead of `TypeError`, `ValueError` etc, which are more appropriate in the main code base. Indeed, you can pass your own message, but the exception class stays the same, which might create some issues in unexpected places.

For example, function A should raise `ValueError` if passed value doesn't fit expected range, and `TypeError` if the input is wrong type. Function B, which calls A, might want to do something specific (eg. cleanup) based on the exception A raises. Raising same `AttributeError` for both different issues will just complicate things (and doesn't go well with *explicit is better than implicit* rule).

Similarly, it renders exception inheritance useless. You might want to catch all `OSError` exceptions, but if all you get is `AssertionError`, try to differentiate between `OSError` sub-exceptions and anything else.

While there isn't any written rule about it, using assert in production code creates more issues than it solves.

> Writing tests checking that a function interface is observed for every invocation of that function doesn't scale. Letting the function check its arguments works better. Whether it's with an assertion, or with an if statement, doesn't really matter.

If the function has nondeterministic interface, then the problem lays much deeper and it should be fixed, not hidden. **However, if you really need to check the attributes (eg. we don't have an init function that sanitize the values), please raise relevant exception (whether it's in `try/except` block or `if condition: raise X`) instead of assertion.**

> That's why we have backtraces :)

Indeed, so instead of "here is the function that messed up" we can go dig through a bunch of them :)

Reply:
I think there are cons and pros for each approach including using `assertion` and `exception`. In my opinion, this PR focus on rework the `merge patch file`. We could probably consider the idea's @spbnick in another issue with another PR in order to have `precondition checking` not only for `merge_patch_file` function, but also for other functions relating to `precondition checking`. ( Of course with the approval of @veruu ). So I mean we keep the raise `exception` in this case in the PR. What do you think?

Reply:
Works for me, what about you @spbnick ? Let's not mix in problems different from what this pull is trying to solve

Reply:
@danghai, could you please address my [first comment](https://github.com/RH-FMK/skt/pull/105/commits/1817bd1bb06c3203c630ed5ecae6a9053011ea1e#r186373587) here? I.e. move calling `os.path.abspath` to `cmd_merge`.